# Feature engineering - missing data

### Import potrzebnych bibliotek

In [87]:
import pandas as pd
import numpy as np

### 1. Wczytaj zbiór danych Titanic (Lab2)

In [88]:
data = pd.read_csv('C:/Users/48692/Desktop/ML/Lab3/data.csv')

### 2. Dokonaj wstępnej inspekcji pierwszych 20 wierszy. Co widzisz?

In [89]:
data.head(20)

pclass  survived                                               name  \
0        1         1                      Allen, Miss. Elisabeth Walton   
1        1         1                     Allison, Master. Hudson Trevor   
2        1         0                       Allison, Miss. Helen Loraine   
3        1         0               Allison, Mr. Hudson Joshua Creighton   
4        1         0    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)   
5        1         1                                Anderson, Mr. Harry   
6        1         1                  Andrews, Miss. Kornelia Theodosia   
7        1         0                             Andrews, Mr. Thomas Jr   
8        1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
9        1         0                            Artagaveytia, Mr. Ramon   
10       1         0                             Astor, Col. John Jacob   
11       1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
12       1         1                      Aubart, Mme. Leontine Pauline   
13       1         1                       Barber, Miss. Ellen 'Nellie'   
14       1         1               Barkworth, Mr. Algernon Henry Wilson   
15       1         0                                Baumann, Mr. John D   
16       1         0                           Baxter, Mr. Quigg Edmond   
17       1         1    Baxter, Mrs. James (Helene DeLaudeniere Chaput)   
18       1         1                              Bazzani, Miss. Albina   
19       1         0                               Beattie, Mr. Thomson   

       sex     age  sibsp  parch    ticket      fare    cabin embarked boat  \
0   female      29      0      0     24160  211.3375       B5        S    2   
1     male  0.9167      1      2    113781    151.55  C22 C26        S   11   
2   female       2      1      2    113781    151.55  C22 C26        S    ?   
3     male      30      1      2    113781    151.55  C22 C26        S    ?   
4   female      25      1      2    113781    151.55  C22 C26        S    ?   
5     male      48      0      0     19952     26.55      E12        S    3   
6   female      63      1      0     13502   77.9583       D7        S   10   
7     male      39      0      0    112050         0      A36        S    ?   
8   female      53      2      0     11769   51.4792     C101        S    D   
9     male      71      0      0  PC 17609   49.5042        ?        C    ?   
10    male      47      1      0  PC 17757   227.525  C62 C64        C    ?   
11  female      18      1      0  PC 17757   227.525  C62 C64        C    4   
12  female      24      0      0  PC 17477      69.3      B35        C    9   
13  female      26      0      0     19877     78.85        ?        S    6   
14    male      80      0      0     27042        30      A23        S    B   
15    male       ?      0      0  PC 17318    25.925        ?        S    ?   
16    male      24      0      1  PC 17558  247.5208  B58 B60        C    ?   
17  female      50      0      1  PC 17558  247.5208  B58 B60        C    6   
18  female      32      0      0     11813   76.2917      D15        C    8   
19    male      36      0      0     13050   75.2417       C6        C    A   

   body                        home.dest  
0     ?                     St Louis, MO  
1     ?  Montreal, PQ / Chesterville, ON  
2     ?  Montreal, PQ / Chesterville, ON  
3   135  Montreal, PQ / Chesterville, ON  
4     ?  Montreal, PQ / Chesterville, ON  
5     ?                     New York, NY  
6     ?                       Hudson, NY  
7     ?                      Belfast, NI  
8     ?              Bayside, Queens, NY  
9    22              Montevideo, Uruguay  
10  124                     New York, NY  
11    ?                     New York, NY  
12    ?                    Paris, France  
13    ?                                ?  
14    ?                    Hessle, Yorks  
15    ?                     New York, NY  
16    ?                     Montreal,

Zaimportowany zestaw danych dotyczy pasażerów uczestniczących w rejsie Titanica. Na podstawie wstępnej inspekcji, można zaobserwować, że zbiór danych składa się z 14 zmiennych (pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body oraz home.dest). Po wyświetleniu pierwszych 20 wierszy, można zaobserwować, że w danych występuje wiele brakujących wartości (oznaczonych znakiem zapytania), w szczególności dla atrybutów boat i body. Dodatkowo dane mogą nie być ujednolicone np. różne zaokrąglenia w przypadku zmiennej age.

### 3. Sprawdź ile jest wartości brakujących używając funkcji pd.isnull().sum() oraz pd.isnull().mean. W sprawozdaniu zapisz swoje obserwacje i skomentuj % brakujących danych w zależności od danej zmiennej.

Przed użyciem powyższych funkcji należało zmodyfikować zbiór danych, poprzez zastąpienie wartości '?' wartością NaN, rozumianą przez interpreter jako reprezentacja braku danych. W tym celu wykorzystano funkcję replace().

In [90]:
data = data.replace('?', np.nan)
pd.isnull(data).sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

In [91]:
pd.isnull(data).mean()

pclass       0.000000
survived     0.000000
name         0.000000
sex          0.000000
age          0.200917
sibsp        0.000000
parch        0.000000
ticket       0.000000
fare         0.000764
cabin        0.774637
embarked     0.001528
boat         0.628724
body         0.907563
home.dest    0.430863
dtype: float64

Wykonanie powyższych operacji pozwoliło stwierdzić, że braki danych występują w przypadku zmiennych age, fare, cabin, embarked, boat, body, home.dest, przy czym najwięcej brakujących wartości zaobserwowano dla atrybutu body, co stanowiło aż około 90.8% wszystkich obserwacji. Wysokim odsetkiem brakujących danych odznaczały się również zmienne cabin i boat (powyżej 60.0%). Niepokojąco dużo brakujących wartości wystąpiło również w przypadku zmiennej home.dest (43.1%) i age (20.1%). Tylko 2 brakujące wartości otrzymano dla atrybutu embarked (0.2%) i jedną dla fare (0.1%).

### 4. Zastanów się z jakim rodzajem braków danych masz do czynienia – całkowicie przypadkowe (MCAR), przypadkowe (MAR) czy nie przypadkowe (MNAR). Swój wybór uzasadnij. Pamiętaj, że są to dane tragicznego rejsu, w którym spora część pasażerów nie przeżyła. Skomentuj wynik.

Utworzenie binarnej zmiennej dla brakujących wartości w kolumnie age:

In [92]:
data['ageNull'] =  np.where(data['age'].isnull(), 1, 0)

Zmapowanie wybranej kolumny ze zmienną survived:

In [93]:
data.groupby(['survived'])['ageNull'].mean()

survived
0    0.234858
1    0.146000
Name: ageNull, dtype: float64

Zapisanie powyższych dwóch linijek kodu za pomocą jednej:

In [94]:
data[['fare','embarked','age','cabin','boat','body','home.dest']].isnull().groupby(data['survived']).mean()

fare  embarked       age     cabin      boat      body  \
survived                                                               
0         0.001236     0.000  0.234858  0.873918  0.988875  0.850433   
1         0.000000     0.004  0.146000  0.614000  0.046000  1.000000   

          home.dest  
survived             
0          0.508035  
1          0.306000

Procentowy udział brakujących wartości dla poszczególnych zmiennych, w zależności czy pasażer Titanica przeżył czy zginął, świadczy o nieprzypadkowych brakach danych (MNAR).

* Nieznaczy odsetek NaN w przypadku zmiennej fare (cena biletu) dla osób, które zginęły, może być wynikiem niedopatrzenia lub podróżnika na gapę.

* Trudno wskazać przyczynę nielicznych braków danych dla zmiennej embarked (port gdzie wsiadł pasażer). Możliwe, że w jakiś sposób zostały po prostu utracone na przestrzeni lat.

* W przypadku zmiennej age (wiek) występuje dużo większy odsetek brakujących danych w obrębie pasażerów, którzy nie przeżyli. Może się to wiązać z trudnością ustalenia dokładnego wieku zmarłych.

* Tak duży odsetek brakujących danych, zarówno w przypadku przeżycia rejsu jak i śmierci pasażera dla zmiennej cabin (numer kabiny), może być związany z nie przykładaniem większej uwagi do tego, jaki był dokładny numer kabiny, szybkiego zapomnienia zbędnych informacji, braku możliwości ustalenia w przypadku zmarłych.

* Prawie 99.0% brakujących wartości w przypadku śmierci pasażera dla zmiennej boat (numer lódki którą się ewakuował), może świadczyć o tym, że prawie wszyscy pasażerowie, którzy zginęli, nie mieli szansy ewakuować się przy pomocy szalupy. Nieliczni dostali szansę, jednak również zginęli. Niewielki % brakujących wartości w przypadku ocalałych, może być wynikiem chaosu, nie zdążeniem zebrania takiej informacji.

* W przypadku zmiennej body (numer identyfikacyjny znalezionego ciała) 85.0%-owy udział brakujących wartości dla przypadku, gdy pasażer nie przeżył może świadczyć o tym, że ciała nie udało się odnaleźć (zostało przeniesione na dalsze odległości, uwięzione we wraku statku itp.), część jednak zostało zidentyfikowanych. Nie było potrzeby wprowadzać danych dla pasażerów, którzy przeżyli.

* Podobnie jak w przypadku zmiennej age, dla zmiennej home.dest występuje dużo większy odsetek brakujących danych w obrębie pasażerów, którzy nie przeżyli rejsu. Może być to spowodowane brakiem chęci podawania dokładnego celu swojej podróży.

### 5. Po powyższej analizie odpowiedz na pytanie w jaki sposób należy postąpić z brakującymi wartościami.

In [95]:
null_sum = data.isnull().sum().sum()
print(null_sum)

3855


W przypadku zmiennej fare oraz embarked, można usunąć brakujące wartości, gdyż stanowią mniej niż 5% zbioru danych. Dla pozostałych zmiennych, w których zaobserwowano braki danych, ich udział wynosi powyżej 5% całości zbioru. W związku z tym brakujące wartości są istotne i nie można ich usunąć. W przypadku gdy rozkład byłby rozkładem normalnym, braki danych mogłyby zostać zastąpiąne wartością średnią lub medianą. W pozostałych przypadkach  można utworzyć nową kategorię "missing".